### PR TITLE
refactor error handling

### PR DESCRIPTION
**Pull request by petersg83**

### CODE REVIEW

File: packages/core/utils/lib/validators.js [hunk 1]
```diff
@@ -56,27 +57,11 @@ class StrapiIDSchema extends MixedSchemaType {
...
-    return acc;
-  }, {});
+const handleYupError = error => {
+  throw new YupValidationError(error);
 };
 
 module.exports = {
   yup,
-  formatYupErrors,
+  handleYupError,
```
Comment (alexandrebodin) [inline packages/core/utils/lib/validators.js:66]: I think we should go one step further and export a validation function directly to wrap all this catching logic

```js
const validateYupSchema = (schema, options = { strict: true, abortEarly: false ) => { 
  return schema.validate(options).catch(err => handleYupError)
}

const validateYupSchemaSync = (schema, options = { strict: true, abortEarly: false ) => { 
  try {
    return schema.validateSync(options)
  } catch (error) {
    handleYupError(error)
  }
}
```

File: test/unit.setup.js [hunk 2]
```diff
@@ -19,10 +19,12 @@ Object.defineProperty(global, 'strapi', {
       plugin.policy = name => plugin.policies[name];
     });
 
-    strapiInstance.service = name => {
+    const oldService = value.service;
```
Comment (alexandrebodin) [inline test/unit.setup.js:22]: when do you need this ?

File: packages/providers/upload-local/lib/index.js [hunk 3]
```diff
@@ -7,13 +7,13 @@
 // Public node modules.
 const fs = require('fs');
 const path = require('path');
-const { errors } = require('@strapi/plugin-upload');
+const { PayloadTooLargeError } = require('@strapi/utils').errors;
```
Comment (alexandrebodin) [inline packages/providers/upload-local/lib/index.js:10]: you can remove @strapi/plugin-upload from the peer dependencies then

File: packages/plugins/users-permissions/server/controllers/validation/user.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  .shape({
+    email: yup
+      .string()
+      .email()
+      .required(),
+    username: yup
+      .string()
+      .min(1)
+      .required(),
+    password: yup
```
Comment (alexandrebodin) [inline packages/plugins/users-permissions/server/controllers/validation/user.js:29]: while at it we can enforce some better default I think

File: packages/plugins/i18n/server/validation/content-types.js [hunk 5]
```diff
@@ -1,9 +1,17 @@
...
+const { YupValidationError } = require('@strapi/utils').errors;
+
 const { get } = require('lodash/fp');
 
-const handleReject = error => Promise.reject(formatYupErrors(error));
+const validateSchema = schema => async (body, errorMessage) => {
+  try {
+    await schema.validate(body, { strict: true, abortEarly: false });
+  } catch (e) {
+    throw new YupValidationError(e, errorMessage);
```
Comment (alexandrebodin) [inline packages/plugins/i18n/server/validation/content-types.js:12]: handleYupError ?

File: packages/plugins/i18n/server/validation/locales.js [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
 const { isoLocales } = require('../constants');
 
 const allowedLocaleCodes = isoLocales.map(prop('code'));
 
-const handleReject = error => Promise.reject(formatYupErrors(error));
+const validateSchema = schema => async (body, errorMessage) => {
+  try {
+    await schema.validate(body, { strict: true, abortEarly: false });
+  } catch (e) {
+    throw new YupValidationError(e, errorMessage);
```
Comment (alexandrebodin) [inline packages/plugins/i18n/server/validation/locales.js:15]: same

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Overall looks really good 👍 

A few notes:

- We can cleanup to do the same handling for all yup error (try / catch once then .catch(..) etc
- I feel like return and empty details object is not ideal and I would rather have the details not returned at all when not necessary

File: packages/core/utils/lib/format-yup-error.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+const { isEmpty } = require('lodash/fp');
+
+// Temporary fix of this issue : https://github.com/jquense/yup/issues/616
+const sanitizeErrorMessage = message =>
+  message.replace(
+    `\n If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\``,
+    ''
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/format-yup-error.js:9]: I saw that, what do you think we can do?

File: packages/core/utils/lib/validators.js (hunk 1 above)
Comment (petersg83) [reply packages/core/utils/lib/validators.js:66]: Yes I was thinking the same

File: packages/core/strapi/package.json [hunk 8]
```diff
@@ -124,7 +124,11 @@
     "package-json": "7.0.0",
     "qs": "6.10.1",
     "resolve-cwd": "3.0.0",
-    "semver": "7.3.5"
+    "semver": "7.3.5",
+    "statuses": "2.0.1"
+  },
+  "devDependencies": {
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/strapi/package.json:130]: I see. It is not necessary but better to have it in the deps. However can't you use the utils from the root test folder

File: packages/core/utils/lib/validators.js (hunk 1 above)
Comment (alexandrebodin) [reply packages/core/utils/lib/validators.js:66]: we can do that a bit later after merging this first pass

File: packages/core/strapi/lib/services/server/koa.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      this.status = 204;
+    } else {
+      this.status = 200;
+      this.body = data;
+    }
+  };
+
+  delegator
+    .method('send')
+    .method('created')
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/strapi/lib/services/server/koa.js:50]: I though the same at the beginning but now it seems ok for me. I think we should keep it that way for the moment

File: packages/core/utils/lib/format-yup-error.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/utils/lib/format-yup-error.js:9]: I'm wondering if we can find a printValue equivalent with util.inspect / json.stringify or sth to make the function simple

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/format-yup-error.js:9]: copy pasting this : https://github.com/jquense/yup/blob/2778b88bdacd5260d593c6468793da2e77daf21f/src/util/printValue.ts + message to origin

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/format-yup-error.js:9]: issue : https://github.com/jquense/yup/blob/2778b88bdacd5260d593c6468793da2e77daf21f/src/util/printValue.ts

File: packages/core/utils/lib/errors.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  constructor(message, details) {
+    super(message, details);
+    this.name = 'ValidationError';
+  }
+}
+
+class YupValidationError extends ValidationError {
+  constructor(yupError, message) {
+    super();
+    const { errors, message: yupMessage } = formatYupErrors(yupError);
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/errors.js:26]: the name is already set by super(). Do you think it should be explicit?

File: packages/core/utils/lib/format-yup-error.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  );
+
+const formatYupInnerError = yupError => ({
+  path: yupError.path
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/format-yup-error.js:13]: I didn't know that function. Nice, thanks

File: test/unit.setup.js (hunk 2 above)
Comment (petersg83) [reply test/unit.setup.js:22]: Well, it doesn't seem needed anymore. With the default value it seems to work at least

File: packages/plugins/users-permissions/server/controllers/validation/user.js (hunk 4 above)
Comment (petersg83) [reply packages/plugins/users-permissions/server/controllers/validation/user.js:20]: I would prefer to focus on rest. Is it ok for you if we make this change later?

File: test/unit.setup.js [hunk 12]
```diff
@@ -8,6 +8,25 @@ Object.defineProperty(global, 'strapi', {
   get() {
     return strapiInstance;
   },
+  // set(value) {
```
Comment (alexandrebodin) [inline test/unit.setup.js:11]: I think you can clean this up

**Review by alexandrebodin — APPROVED** [review]
This is a big first step 👍

**Comment by derrickmehaffy** [comment]
Woooo! Yayyyy!

**Pull request merged**